tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreManagedLedgerOffloader.java

```diff
@@ -483,59 +227,37 @@ public String getOffloadDriverName() {
         return promise;
     }
 
-    String getReadRegion(Map<String, String> offloadDriverMetadata) {
-        return offloadDriverMetadata.getOrDefault(METADATA_FIELD_REGION, writeRegion);
-    }
-
-    String getReadBucket(Map<String, String> offloadDriverMetadata) {
-        return offloadDriverMetadata.getOrDefault(METADATA_FIELD_BUCKET, writeBucket);
-    }
-
-    String getReadEndpoint(Map<String, String> offloadDriverMetadata) {
-        return offloadDriverMetadata.getOrDefault(METADATA_FIELD_ENDPOINT, writeEndpoint);
-    }
-
-    BlobStore getReadBlobStore(Map<String, String> offloadDriverMetadata) {
-        BlobStoreLocation location = BlobStoreLocation.of(
-            getReadRegion(offloadDriverMetadata),
-            getReadEndpoint(offloadDriverMetadata)
-        );
-        BlobStore blobStore = readBlobStores.get(location);
-        if (null == blobStore) {
-            blobStore = createBlobStore(
-                offloadDriverName,
-                location.getRegion(),
-                location.getEndpoint(),
-                credentials,
-                maxBlockSize
-            ).getRight();
-            BlobStore existingBlobStore = readBlobStores.putIfAbsent(location, blobStore);
-            if (null == existingBlobStore) {
-                return blobStore;
-            } else {
-                return existingBlobStore;
-            }
-        } else {
-            return blobStore;
-        }
+    /**
+     * Attempts to create a BlobStoreLocation from the values in the offloadDriverMetadata,
+     * however, if no values are available, it defaults to the currently configured
+     * provider, region, bucket, etc.
+     *
+     * @param offloadDriverMetadata
+     * @return
+     */
+    private BlobStoreLocation getBlobStoreLocation(Map<String, String> offloadDriverMetadata) {
+        return (!offloadDriverMetadata.isEmpty()) ? new BlobStoreLocation(offloadDriverMetadata) :
+            new BlobStoreLocation(getOffloadDriverMetadata());
     }
 
     @Override
     public CompletableFuture<ReadHandle> readOffloaded(long ledgerId, UUID uid,
                                                        Map<String, String> offloadDriverMetadata) {
-        String readBucket = getReadBucket(offloadDriverMetadata);
-        BlobStore readBlobstore = getReadBlobStore(offloadDriverMetadata);
+
+        BlobStoreLocation bsKey = getBlobStoreLocation(offloadDriverMetadata);
+        String readBucket = bsKey.getBucket();
+        BlobStore readBlobstore = blobStore;
```

Comment:
readBlobStore may not necessarily be the same as the writeBlobStore, in the case that configuration has changed. This is why they were differentiated before. Provider, region & endpoint can change

Reply:
That's why I had to the BlobStoreRepository class to keep these, and be able to look them up based on the provider, region, endpoint parameters passed in.

There is a similar map of these bookstores in the current version of the BlobStoreManagedLedgerOffloader class I am refactoring.

private final ConcurrentMap<BlobStoreLocation, BlobStore> readBlobStores = new ConcurrentHashMap<>(); 

If I cannot keep these blobstore instances somewhere, then I cannot fix any of the unit tests. So if removing the BlobStoreRepository is a MUST have, then let's just kill this PR and keep the current implementation, as I don't see any other way of fixing the unit tests.

Reply:
My issue with BlobStoreRepository was that it was global. If you can scope it within the BlobStoreManagedLedgerOffloader or pass it in as a parameter, that removes the issue with it.

Which unit tests can't run without the BlobStoreRepository?

Reply:
I will limit the scope of the BlobStoreRepository.  Without all of the unit tests that interact with the blobstore more than once, i.e write, then read to verify the results, fail. This is due to the fact that for the unit tests I use the transient blog store, so every invocation of the getBlobStore() method creates a new instance, which doesn't have the data that was written to the previous blogStore.  This isn't an issue for AWS, etc blog stores, as the same configuration points to the same location on S3.

Reply:
This unit test in particular are you referring to though?

Reply:
@ivankelly There were 6 tests in the BlobStoreManagedLedgerOffloaderTest class that failed.

testOffloadAndRead()
testOffloadFailInitDataBlockUpload()
testOffloadFailDataBlockPartUpload()
testDeleteOffloaded()
testDeleteOffloadedFail()
testReadUnknownDataVersion()

But I have since fixed them.  The Integration tests are failing due to the test container failing to open a specific port.